vertical-pod-autoscaler/initializer/initializer.go

```diff
@@ -0,0 +1,347 @@
+	}
+
+}
+
+func getFailedPod(pod *v1.Pod) (*v1.Pod, error) {
+	failedPodCopy, err := api.Scheme.Copy(pod)
+	if err != nil {
+		return nil, err
+	}
+	failedPod := failedPodCopy.(*v1.Pod)
+	markInitializationFailure(failedPod)
+	return failedPod, nil
+}
+
+// getInitializedPod returns a pod in an initialized state, updating resource
+// requests if applicable.
+func (i *initializer) getInitializedPod(pod *v1.Pod) (*v1.Pod, error) {
+
+	glog.V(4).Infof("Updating requirements for pod: %v", pod.Name)
+
+	updatedPodCopy, err := api.Scheme.Copy(pod)
+	if err != nil {
+		return nil, err
+	}
+	updatedPod := updatedPodCopy.(*v1.Pod)
+	glog.Infof("UPDATED POD: %+v", updatedPod)
+	markInitializationSuccess(updatedPod)
+
+	vpaConfig := i.getMatchingVPA(pod)
+	if vpaConfig == nil {
+		glog.V(4).Infof("No matching vpa found.")
+		return updatedPod, nil
+	}
+
+	recommendation, err := i.recommender.Get(&pod.Spec)
+	if err != nil || recommendation == nil {
+		if vpaConfig.Status.Recommendation != nil {
+			// fallback to recommendation cached in VPA config
+			recommendation = vpaConfig.Status.Recommendation
+		} else {
+			// no recommendation to apply
+			return updatedPod, nil
+		}
+	}
+
+	glog.V(4).Infof("Recommendation for pod %v: %+v", pod.Name, recommendation)
+
+	i.applyRecomendedResources(updatedPod, recommendation, vpaConfig.Spec.ResourcesPolicy)
+
+	return updatedPod, nil
+}
+
+func shouldInitialize(pod *v1.Pod) bool {
+	return pod.ObjectMeta.Initializers != nil && len(pod.ObjectMeta.Initializers.Pending) > 0 && pod.ObjectMeta.Initializers.Pending[0].Name == vpaInitializerName
+}
+
+// markInitializationSuccess denotes successfull initalization for pod
+func markInitializationSuccess(pod *v1.Pod) {
+	if len(pod.ObjectMeta.Initializers.Pending) == 1 {
+		pod.ObjectMeta.Initializers = nil
+	} else {
+		pod.ObjectMeta.Initializers.Pending = pod.ObjectMeta.Initializers.Pending[1:]
+	}
+}
+
+// markInitializationFailure denotes failed initalization for pod
+func markInitializationFailure(pod *v1.Pod) {
+	pod.ObjectMeta.Initializers.Result = &metav1.Status{Status: metav1.StatusFailure}
+}
+
+// applyRecomendedResources overwrites pod resources Request field with recommended values.
+func (i *initializer) applyRecomendedResources(pod *v1.Pod, recommendation *apimock.Recommendation, policy apimock.ResourcesPolicy) {
+	for _, container := range pod.Spec.Containers {
+		containerRecommendation := getRecommendationForContainer(recommendation, container)
+		if containerRecommendation == nil {
+			continue
+		}
+		containerPolicy := getContainerPolicy(container.Name, &policy)
+		applyPolicy(containerRecommendation, containerPolicy)
+		for resource, recommended := range containerRecommendation.Resources {
+			requested, exists := container.Resources.Requests[resource]
+			if exists {
+				// overwriting existing resource spec
+				glog.V(2).Infof("updating resources request for pod %v container %v resource %v old value %v new value %v",
```

Comment:
Please define what is V(1), V(2) V(4) - i see that there are some inconsistencies in logging levels.